### PR TITLE
Introduce FloatingPointCounter

### PR DESCRIPTION
**Pull request by rauhul**
Introduce FloatingPointCounter.

### Motivation:

It is not currently possible to record floating point values via the swift-metrics API even if the metrics backend supports it.

### Modifications:

Adds a `FloatingPointCounter` type to allow users to accumulate non-integral metrics backed by a `FloatingPointCounterHandler`. Introduces a default implementation for creating and destroying `FloatingPointCounterHandler`s for metric backends that do not natively support floating point counters.

On such backends, `FloatingPointCounter` is backed by a `AccumulatingRoundingFloatingPointCounter` which accumulates floating point values internally and record increments to a wrapped `CounterHandler` after crossing integer boundaries.

### Result:

Users can create `FloatingPointCounter`s to record floating point values and get enhanced behavior for backends that support floating point values.

### CODE REVIEW

File: Sources/CoreMetrics/Metrics.swift [hunk 1]
```diff
@@ -488,6 +580,62 @@ public protocol MetricsFactory {
...
+    var rawValue: Double = 0
+    var previousValue: Int64 = 0
+    var counterHandler: CounterHandler
+
+    init(counterHandler: CounterHandler) {
+        self.counterHandler = counterHandler
+    }
+
+    func increment(by amount: Double) {
+        // FIXME: does this need a lock?
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:594]: or use an atomic

Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift:594]: Wouldn't the entire function body need to be atomic, otherwise concurrent increments could mess-up the internal state

File: Sources/CoreMetrics/Metrics.swift [hunk 2]
```diff
@@ -488,6 +580,62 @@ public protocol MetricsFactory {
...
+        rawValue += amount
+        let increment = Int64(rawValue) - previousValue
+        assert(increment >= 0)
+        if increment > 0 {
+            previousValue += increment
+            counterHandler.increment(by: increment)
+        }
+    }
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:602]: we should protect from overflows ^^

**Comment by tomerd** [comment]
@swift-server-bot add to allowlist

**Comment by rauhul** [comment]
Should these types be called `FloatCounter*`?

File: Sources/CoreMetrics/Metrics.swift [hunk 3]
```diff
@@ -488,6 +580,74 @@ public protocol MetricsFactory {
...
+    /// Invoked when the corresponding `FloatingPointCounter`'s `destroy()` function is invoked.
+    /// Upon receiving this signal the factory may eagerly release any resources related to this counter.
+    ///
+    /// `destroyFloatingPointCounter` must be implemented if `makeFloatingPointCounter` is implemented.
+    ///
+    /// - parameters:
+    ///     - handler: The handler to be destroyed.
+    public func destroyFloatingPointCounter(_ handler: FloatingPointCounterHandler) {
+        if let handler = handler as? AccumulatingRoundingFloatingPointCounter {
+            destroyCounter(handler.counterHandler)
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:646]: handler.destroy and then handler.counterHandler can stay private?

File: Sources/CoreMetrics/Metrics.swift [hunk 4]
```diff
@@ -488,6 +580,74 @@ public protocol MetricsFactory {
...
+                self.counterHandler.increment(by: roundedIncrement)
+            }
+        }
+    }
+
+    func reset() {
+        self.lock.withLockVoid {
+            self.rawValue = 0
+            self.roundedValue = 0
+            self.counterHandler.reset()
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:619]: self.counterHandler.reset() outside of the lock?

File: Sources/CoreMetrics/Metrics.swift [hunk 5]
```diff
@@ -488,6 +580,74 @@ public protocol MetricsFactory {
...
 }
 
+/// Wraps a CounterHandler, adding support for incrementing by floating point values by storing an accumulated floating point value and recoding increments to the underlying CounterHandler after crossing integer boundaries.
+private class AccumulatingRoundingFloatingPointCounter: FloatingPointCounterHandler {
+    private static let kAutoStateResetThreshold = Double(Int64.max - (2 << 8))
+
+    private var lock = Lock()
+    private var rawValue: Double = 0
+    private var roundedValue: Int64 = 0
+    private(set) var counterHandler: CounterHandler
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:590]: can this be a let?

File: Sources/CoreMetrics/Metrics.swift [hunk 6]
```diff
@@ -488,6 +580,74 @@ public protocol MetricsFactory {
     func destroyTimer(_ handler: TimerHandler)
 }
 
+/// Wraps a CounterHandler, adding support for incrementing by floating point values by storing an accumulated floating point value and recoding increments to the underlying CounterHandler after crossing integer boundaries.
+private class AccumulatingRoundingFloatingPointCounter: FloatingPointCounterHandler {
+    private static let kAutoStateResetThreshold = Double(Int64.max - (2 << 8))
+
+    private var lock = Lock()
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:587]: can this be a let?

File: Sources/CoreMetrics/Metrics.swift (hunk 4 above)
Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift:619]: That could interleave with calls to `increment` though unlikely, right?

File: Sources/CoreMetrics/Metrics.swift [hunk 7]
```diff
@@ -488,6 +580,74 @@ public protocol MetricsFactory {
...
+    func increment(by amount: Double) {
+        assert(amount < AccumulatingRoundingFloatingPointCounter.kAutoStateResetThreshold)
+        self.lock.withLockVoid {
+            let incremented = rawValue + amount
+            if incremented < AccumulatingRoundingFloatingPointCounter.kAutoStateResetThreshold {
+                self.rawValue = incremented
+            } else {
+                self.rawValue = amount
+                self.roundedValue = 0
+            }
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:605]: in other places (e.g. Timer) we compare to Int.max and continue to emit max instead of rolling to 0. not sure what a better approach in this case.

Comment (tomerd) [reply Sources/CoreMetrics/Metrics.swift:605]: also we should add unit test to test normal + overflow behaviors

Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift:605]: this defers that behavior to the underlying `CounterHandler` and just ensures that the computation in `increment` won't trap or overflow. though looking at this closer, I think the logic is wrong, hmm...

File: Sources/CoreMetrics/Metrics.swift (hunk 4 above)
Comment (tomerd) [reply Sources/CoreMetrics/Metrics.swift:619]: my is worry about a deadlock if the handler gets stuck for whatever reason. presumably, the handler can handle its own concurrency so we dont need to protect it?

File: Sources/CoreMetrics/Metrics.swift (hunk 3 above)
Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift]: Are you thinking I should add a destroy method? The destroy instance funcs are only available on the `<Metric>` API not the `<Metrics>Handler` SPI, iiuc.

Comment (tomerd) [reply Sources/CoreMetrics/Metrics.swift]: right, since AccumulatingRoundingFloatingPointCounter is an internal type and we check the type here, we can presumably add a non-standard func to it?

Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift]: Well it will need access to the metrics factory so the method would end up looking like:
```swift
    func destroy(factory: MetricsFactory) {
        factory.destroyCounter(counterHandler)
    }

    public func destroyFloatingPointCounter(_ handler: FloatingPointCounterHandler) {
        if let handler = handler as? AccumulatingRoundingFloatingPointCounter {
            handler.destroy(factory: self)
        }
    }
```
I wasn't sure if we should pass the metrics factory to `AccumulatingRoundingFloatingPointCounter`; I'm happy to make this change if you think it makes sense.

Comment (tomerd) [reply Sources/CoreMetrics/Metrics.swift]: factory is sort of a s singleton, so maybe `MetricsSystem.factory.destroyCounter(self.counterHandler)` would work? none of them are too elegant, but at least with this the private state of FloatingPointCounterHandler is kept private

Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift]: Pushed this change

File: Sources/CoreMetrics/Metrics.swift [hunk 8]
```diff
@@ -488,6 +580,76 @@ public protocol MetricsFactory {
...
+    /// Create a default backing `FloatingPointCounterHandler` for backends which do not naively support floating point counters.
+    ///
+    /// The created FloatingPointCounterHandler is a wrapper around a backend's CounterHandler which accumulates floating point increments and records increments to an underlying CounterHandler after crossing integer boundaries.
+    ///
+    /// - parameters:
+    ///     - label: The label for the `FloatingPointCounterHandler`.
+    ///     - dimensions: The dimensions for the `FloatingPointCounterHandler`.
+    public func makeFloatingPointCounter(label: String, dimensions: [(String, String)]) -> FloatingPointCounterHandler {
+        return AccumulatingRoundingFloatingPointCounter(
+            counterHandler: self.makeCounter(label: label, dimensions: dimensions))
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:638]: I guess we can also simplify this by moving the counter handler creation into the ctor of AccumulatingRoundingFloatingPointCounter

e.g.

```
 public func makeFloatingPointCounter(label: String, dimensions: [(String, String)]) -> FloatingPointCounterHandler {
  return AccumulatingRoundingFloatingPointCounter(label: label, dimensions: dimensions)
}

private class AccumulatingRoundingFloatingPointCounter: FloatingPointCounterHandler {
     private let counterHandler: CounterHandler

     init(label: String dimension: [String: String]) {
         self.counterHandler = MetricsSystem.factory.makeCounter(label: label, dimension: dimension)
     }
```

File: Sources/CoreMetrics/Metrics.swift [hunk 9]
```diff
@@ -488,6 +580,76 @@ public protocol MetricsFactory {
...
+
+    init(counterHandler: CounterHandler) {
+        self.counterHandler = counterHandler
+    }
+
+    func increment(by amount: Double) {
+        // FIXME: what do?
+        let maxIncrement = Double(Int64.max - (2 << 8))
+        assert((0...maxIncrement).contains(amount))
+        let amount = min(amount, maxIncrement)
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:597]: not exactly sure I understand what the above is validating, its it that amount < Int64.max?

Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift:597]: I wanted to make sure that doubles passed to this integer rounded counter were representable, the newer version would look like this:
```swift
        let largestRepresentableInt64InFloat64 = Double(2 << 53)
        assert((0...largestRepresentableInt64InFloat64).contains(amount))
        let amount = min(amount, largestRepresentableInt64InFloat64)
```

I'm not sure if this is the right behavior though.

File: Sources/CoreMetrics/Metrics.swift (hunk 4 above)
Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift:619]: I don't think that is thread safe:

``` swift
 Time | Thread A                 | Thread B
------+--------------------------+-------------------
    0 | counter.inc(...)         | counter.reset()
    1 | <blocked>                | lock
    2 | <blocked>                | fraction = 0
    3 | <blocked>                | unlock
    4 | lock                     | <not scheduled>
    5 | fraction = x             | <not scheduled>
    6 | handler.inc(...)         | <not scheduled>
    7 | unlock                   | <not scheduled>
    8 | return                   | handler.reset()
    9 |                          | return
```

Depending on the value passed to `inc` and the value in `fraction` you could get incorrect results I think.

Comment (tomerd) [reply Sources/CoreMetrics/Metrics.swift:619]: oh, I think you are worried about correctness, not actual concurrent access, right?

Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift:619]: Yep, it seems like a correctness issue to not lock around the reset

File: Sources/CoreMetrics/Metrics.swift (hunk 9 above)
Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift]: This behavior changed, we should probably start a new thread

**Comment by rauhul** [comment]
cc @weissi @ktoso @stephentyrone
thoughts?

**Comment by weissi** [comment]
CC @MrLotU

**Comment by MrLotU** [comment]
Looks like a great feature to have! I just opened https://github.com/MrLotU/SwiftPrometheus/pull/55 as a companion, adding native support to SwiftPrometheus 😄

**Comment by tomerd** [comment]
@rauhul lets add a few tests?

**Comment by rauhul** [comment]
> @rauhul lets add a few tests?

@tomerd just pushed some tests

File: Sources/CoreMetrics/Metrics.swift [hunk 10]
```diff
@@ -12,23 +12,27 @@
...
 //===----------------------------------------------------------------------===//
 
+// MARK: Testing API
+
+internal var _enableAssertions = true
+
 // MARK: User API
 
-extension Counter {
+public extension Counter {
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:21]: these formatting changes are not desired. we prefer to make the methods public not the extension. what version of swift-format did you use?

File: Sources/CoreMetrics/Metrics.swift [hunk 11]
```diff
@@ -12,23 +12,27 @@
 //
 //===----------------------------------------------------------------------===//
 
+// MARK: Testing API
+
+internal var _enableAssertions = true
```
Comment (tomerd) [inline Sources/CoreMetrics/Metrics.swift:17]: not sure how I feel about this flag. can we get rid of it?

File: Sources/CoreMetrics/Metrics.swift (hunk 10 above)
Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift:21]: `0.48.4`

File: Sources/CoreMetrics/Metrics.swift (hunk 11 above)
Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift:17]: I don't see a way to test the assertion path without it, do you have a better solution?

File: Tests/MetricsTests/CoreMetricsTests.swift [hunk 12]
```diff
@@ -16,10 +16,19 @@
 import XCTest
 
 class MetricsTests: XCTestCase {
-    func testCounters() throws {
+    let metrics = TestMetrics()
```
Comment (tomerd) [inline Tests/MetricsTests/CoreMetricsTests.swift:19]: prefer little global state as possible

File: Tests/MetricsTests/MetricsTests.swift [hunk 13]
```diff
@@ -17,26 +17,25 @@
 import XCTest
 
 class MetricsExtensionsTests: XCTestCase {
-    func testTimerBlock() throws {
+    let metrics = TestMetrics()
```
Comment (tomerd) [inline Tests/MetricsTests/MetricsTests.swift:20]: ditto re global state

File: Tests/MetricsTests/TestMetrics.swift [hunk 14]
```diff
@@ -154,7 +154,7 @@ internal class TestTimer: TimerHandler, Equatable {
         }
     }
 
-    func retriveValueInPreferredUnit(atIndex i: Int) -> Double {
+    func retrieveValueInPreferredUnit(atIndex i: Int) -> Double {
```
Comment (tomerd) [inline Tests/MetricsTests/TestMetrics.swift:157]: <3

File: Tests/MetricsTests/MetricsTests.swift (hunk 13 above)
Comment (rauhul) [reply Tests/MetricsTests/MetricsTests.swift:20]: I can revert these

File: Sources/CoreMetrics/Metrics.swift (hunk 10 above)
Comment (tomerd) [reply Sources/CoreMetrics/Metrics.swift:21]: sorry about that :/ can you revert these and use 0.44.6? one swiftPM plugins are there, this kind of issues will go away hopefully.

Comment (tomerd) [reply Sources/CoreMetrics/Metrics.swift:21]: or if you want to keep the local version - run it via the docker setup

File: Sources/CoreMetrics/Metrics.swift (hunk 11 above)
Comment (tomerd) [reply Sources/CoreMetrics/Metrics.swift:17]: dunno! @weissi @ktoso do you have any? maybe we remove the asserts?

File: Tests/MetricsTests/MetricsTests.swift (hunk 13 above)
Comment (rauhul) [reply Tests/MetricsTests/MetricsTests.swift]: fixed in `ae6a0cd`

File: Sources/CoreMetrics/Metrics.swift (hunk 11 above)
Comment (weissi) [reply Sources/CoreMetrics/Metrics.swift:17]: In NIO we test them in [`NIOCrashTester`](https://github.com/apple/swift-nio/blob/main/Sources/NIOCrashTester/CrashTests%2BEventLoop.swift).

Comment (ktoso) [reply Sources/CoreMetrics/Metrics.swift:17]: I used to have a big test suite for crashes but we removed that a while ago. Nowadays doing so in integration tests only. 
CrashTester sounds like the way to go -- we could "steal" that type I guess.

Comment (weissi) [reply Sources/CoreMetrics/Metrics.swift:17]: yeah, go ahead. Just need to copy the three "driver" files and for every new test suite, you'll need to put an entry into this [dict](https://github.com/apple/swift-nio/blob/main/Sources/NIOCrashTester/CrashTestSuites.swift). It's able to automatically enumerate all `test*` methods, so no need to declare every test. (Here's the [method enumeration code](https://github.com/apple/swift-nio/blob/74b3a8d6a09281ea694ab49a3502012f6c5db99d/Sources/NIOCrashTester/main.swift#L76-L89))

File: Sources/CoreMetrics/Metrics.swift (hunk 10 above)
Comment (rauhul) [reply Sources/CoreMetrics/Metrics.swift]: done, ran with `0.44.6`

**Comment by tomerd** [comment]
this feels ready to me. @ktoso @weissi?

**Review by ktoso — COMMENTED** [review]
Sorry I was slow to review here, read through and looks solid, thanks a lot @rauhul !

**Comment by ktoso** [comment]
Double checking the API break:

```
20:09:15 Func MetricsFactory.destroyFloatingPointCounter(_:) has been added as a protocol requirement
20:09:15 Func MetricsFactory.makeFloatingPointCounter(label:dimensions:) has been added as a protocol requirement
```

this is fine because we're also adding implementations along with it:

```
extension MetricsFactory {
    public func makeFloatingPointCounter(label: String, dimensions: [(String, String)]) -> FloatingPointCounterHandler {
```

so that error we can ignore.

**Review by ktoso — APPROVED** [review]

**Comment by tomerd** [comment]
@weissi any thoughts?

**Review by tomerd — APPROVED** [review]

**Comment by ktoso** [comment]
I think we're good here, I'll catch up with Johannes if necessary offline.

**Review by weissi — COMMENTED** [review]
late LGTM. Definitely not an expert here but looks good

**Comment by ktoso** [comment]
Released as 2.2.0 https://github.com/apple/swift-metrics/releases/tag/2.2.0 FYI @MrLotU 👍

**Comment by MrLotU** [comment]
Native SwiftPrometheus support added in https://github.com/MrLotU/SwiftPrometheus/releases/tag/1.0.0-alpha.14
